Change system table check logic

makssent opened the pull request:
Fixes #36462

Changes proposed in this pull request:
- Moved system-table detection to SystemTableManager and relocated it under shardingsphere/database/core.  
- Introduced a dialect-driven SPI: DialectSystemTableManager to provide per-dialect system table lists and checks.  
- Updated SchemaMetaDataLoader to delegate to the dialect implementation’s isSystemTable, falling back to the legacy heuristic when no dialect is available.  
- Added unit tests.  

---

Before committing this PR, I'm sure that I have checked the following options:
- [X] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [X] I have self-reviewed the commit code.
- [X] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.
- [ ] I have updated the Release Notes of the current development version. For more details, see [Update Release Note](https://shardingsphere.apache.org/community/en/involved/contribute/contributor/)

makssent committed:
Change system table check logic in SchemaMetaDataLoader

makssent committed:
fix test

makssent commented:
There are two points I’m not fully satisfied with:

1. In `SystemTableManagerTest` I had to implement a custom schema loading mechanism, because the schemas are not available during tests before the program starts. At runtime everything works correctly, the issue only exists in the test environment. I couldn’t find another solution: adding `dialect-connector-*` dependencies would be a more natural option, but it breaks the build.

2. The TODOs I left. I think they should probably be removed, but I wanted to highlight one point: in `SimpleTableSegmentBinder` we currently use `SystemTableManager.isSystemTable()` without `databaseType`, even though `databaseType` is available and this would not change the program logic. The code could be rewritten like this, and the implementation of `isSystemTable` without `databaseType` could be removed from `SystemTableManager`:

    ```java
    if (SystemTableManager.isSystemTable(binderContext.getSqlStatement().getDatabaseType().getType(),
            schemaName, tableName)) {
        return;
    }
    ```

    This would make the method that only accepts `schema` unnecessary, since it’s not used anywhere else. I didn’t remove it in this PR — maybe you have future plans for it — but I wanted to draw attention to this point.

I verified it with tests. I also checked Firebird manually — everything works as intended.

Looking forward to your review and suggestions for improvement.

strongduanmu commented on database/connector/core/src/main/java/org/apache/shardingsphere/database/connector/core/metadata/manager/SystemTableManager.java:
```diff
@@ -0,0 +1,122 @@
+     * @param tableName    table name
+     * @return whether current table is system table or not
+     */
+    public static boolean isSystemTable(final String databaseType, final String schema, final String tableName) {
```
If this is the case, perhaps we can refactor it without SPI, as long as we ensure that the database without configuring the system table YAML file can go to the default logic.

strongduanmu commented on database/connector/core/src/test/java/org/apache/shardingsphere/database/connector/core/metadata/manager/SystemTableManagerTest.java:
```diff
@@ -0,0 +1,120 @@
+    static void setUp() throws Exception {
+        originalClassLoader = Thread.currentThread().getContextClassLoader();
+        URL[] urls = {
+                Paths.get("../dialect/mysql/src/main/resources").toUri().toURL(),
```
You can write some FIXTUREs and test configurations specifically for testing to test the logic of SystemTableManager without actually loading the dialect module.

strongduanmu commented on database/connector/dialect/firebird/src/main/java/org/apache/shardingsphere/database/connector/firebird/metadata/manager/FirebirdSystemTableManager.java:
```diff
@@ -0,0 +1,31 @@
+public final class FirebirdSystemTableManager implements DialectSystemTableManager {
+    
+    @Override
+    public String getDatabaseType() {
```
It seems that the implementation classes of these SPIs are empty. Based on this, I suggest removing the SPI interface and implementation class and encapsulating these logics through a SystemTableDetector.

strongduanmu reviewed (changes requested):
Finally, I would like to remind you that to determine whether a table is a true system table, you also need to look at the path to which the table belongs (db - schema - table). If the path belongs to a user-defined schema, perhaps this table should be identified as a non-system table.

makssent commented:
Hi, sorry for the late reply — I’m working on another task in parallel. I’ve reviewed your comments and will apply the fixes. Thanks!

makssent committed:
add fixtures for SystemTableManager without dialect loading

makssent committed:
eliminate unnecessary SPI from SystemTableManager

makssent committed:
add Engine to hide SPI loading for SystemTableManager

makssent committed:
rename *Manager to *Detector,
eliminate redundant isSystemTable overload,
remove unused getTables method,
rework SPI from Manager to Rule

makssent committed:
Merge branch 'master' into master_226041_system_tables_filter_fix

makssent commented:
I hope I’ve finished and done everything correctly. I think my problem was also that I just copied the old implementation from `SystemSchemaManager` and didn’t even try to change it, trying not to deviate too much from the base code, not assuming that it was already outdated and not fully aligned with current standards.

I reviewed all the comments and approaches, applied them, and tried to bring the code up to date. I created a new SPI — `SystemTableRule` (now optional, so each database can define its own rules). I removed the unnecessary override of `isSystemTable()` because everywhere it is used we already know the `databaseType`. I also deleted the unused `getTables()` function, renamed `*Manager` to `*Detector`, and did some refactoring. I think it will be better to implement specific logic later, when it is actually needed, instead of keeping it around without use.

I’m looking forward to your next comments, and I hope I didn’t miss anything. It was interesting for me to dig into the core.

makssent commented:
I see that the build is failing, but I noticed the same error appears in other PRs as well. I’ll wait until the issue is resolved on your side and then rerun the CI.

makssent committed:
ci
